In [81]:
from fastai.text.all import *

In [82]:
df = pd.read_csv('./expenses.csv')
df.head()

description                     category
0  TRADER JOE S #068 Merchandise             Groceries & Food
1   WHOLEFDS SVL#10707 Groceries             Groceries & Food
2            Amazon.com Shopping                       Amazon
3                OPENAI Shopping  Subscriptions & Memberships
4  Amazon.com*NH99Y5J73 Shopping                       Amazon

In [91]:
dls = TextDataLoaders.from_csv(
    path='.',
    csv_fname='expenses.csv',
    text_col='description',
    label_col='category')
dls.show_batch(max_n=3)

In [92]:
train_size = len(dls.train_ds)
valid_size = len(dls.valid_ds)

print(f"Training size: {train_size}")
print(f"Validation size: {valid_size}")

Training size: 134
Validation size: 33


# Fine tuning

In [103]:
learn_finetune = text_classifier_learner(dls, AWD_LSTM, drop_mult=0.5, metrics=accuracy)

In [104]:
learn_finetune.fine_tune(10, 1e-2)

In [105]:
learn_finetune.show_results()

In [106]:
learn_finetune.export('models/120523-finetune.pkl')

In [97]:
learn_finetune.predict('AMZN Mktp US*GW42262P3')

('Amazon',
 tensor(0),
 tensor([0.1829, 0.0650, 0.0541, 0.0701, 0.0436, 0.0691, 0.0597, 0.0631, 0.0814,
         0.0523, 0.0583, 0.0632, 0.0651, 0.0720]))

# Fit one cycle

In [98]:
learn_fit = text_classifier_learner(dls, AWD_LSTM, drop_mult=0.5, metrics=accuracy)

In [107]:
learn_fit.fit_one_cycle(10)

In [108]:
learn_fit.show_results()

In [101]:
learn_fit.export('models/120523-fit.pkl')

In [102]:
learn_fit.predict('AMZN Mktp US*GW42262P3')

('Miscellaneous',
 tensor(7),
 tensor([0.0776, 0.0749, 0.0646, 0.0753, 0.0752, 0.0704, 0.0697, 0.0807, 0.0729,
         0.0760, 0.0605, 0.0736, 0.0666, 0.0621]))

# Model Training Results

## 12-05-2023

### Fine tuning

Training size: 134
Validation size: 33

| epoch | train_loss | valid_loss | accuracy | time  |
|-------|------------|------------|----------|-------|
| 0     | 2.804887   | 2.517872   | 0.333333 | 00:00 |
| 0     | 2.155915   | 2.453891   | 0.424242 | 00:00 |
| 1     | 2.023252   | 2.341634   | 0.515152 | 00:00 |
| 2     | 1.836782   | 2.136570   | 0.636364 | 00:00 |
| 3     | 1.706438   | 1.865172   | 0.727273 | 00:00 |
| 4     | 1.591477   | 1.587878   | 0.757576 | 00:00 |
| 5     | 1.477993   | 1.415652   | 0.787879 | 00:00 |
| 6     | 1.391266   | 1.315022   | 0.787879 | 00:00 |
| 7     | 1.316200   | 1.265274   | 0.787879 | 00:00 |
| 8     | 1.248108   | 1.236477   | 0.787879 | 00:00 |
| 9     | 1.189136   | 1.216803   | 0.787879 | 00:00 |


### Fit one cycle

Training size: 134
Validation size: 33

| epoch | train_loss | valid_loss | accuracy | time  |
|-------|------------|------------|----------|-------|
| 0     | 2.399887   | 2.523246   | 0.242424 | 00:00 |
| 1     | 2.380246   | 2.472462   | 0.363636 | 00:00 |
| 2     | 2.319030   | 2.388698   | 0.454545 | 00:00 |
| 3     | 2.246946   | 2.302907   | 0.545455 | 00:00 |
| 4     | 2.165878   | 2.222806   | 0.575758 | 00:00 |
| 5     | 2.109444   | 2.143269   | 0.606061 | 00:00 |
| 6     | 2.055223   | 2.077895   | 0.606061 | 00:00 |
| 7     | 2.002399   | 2.026833   | 0.606061 | 00:00 |
| 8     | 1.979071   | 1.984128   | 0.636364 | 00:00 |
| 9     | 1.940246   | 1.944949   | 0.666667 | 00:00 |



# References
1. [Text transfer learning](https://docs.fast.ai/tutorial.text.html)
2. [TextDataLoaders.from_csv](https://docs.fast.ai/text.data.html#textdataloaders.from_csv)